File: feeder/api/models/feeder.py
from typing import Optional, List, Any
from pydantic import BaseModel, ConfigDict
from feeder.api.models import BasePaginatedList


class GenericResponse(BaseModel):
    success: str = "ok"


class TriggerFeeding(BaseModel):
    portion: float = 0.0625


class FeedEvent(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    device_name: Optional[str] = None
    device_hid: str
    timestamp: int
    start_time: int
    end_time: int
    pour: Optional[int] = None
    full: Optional[int] = None
    grams_expected: int
    grams_actual: int
    hopper_start: int
    hopper_end: int
    source: int
    fail: bool
    trip: Optional[bool] = None
    lrg: Optional[bool] = None
    vol: Optional[bool] = None
    bowl: Optional[bool] = None
    recipe_id: str
    error: Optional[str] = None


class FeedHistory(BasePaginatedList):
    data: List[FeedEvent]


class HopperLevel(BaseModel):
    level: int


class Recipe(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: Optional[int] = None
    name: Optional[str] = ""
    tbsp_per_feeding: Optional[int] = None
    g_per_tbsp: Optional[int] = None
    budget_tbsp: Optional[int] = None


class RawMQTTMessage(BaseModel):
    command: str
    args: Any = {}

File: static/src/shapes/feeder.js
import PropTypes from "prop-types";

export const feederDeviceShape = PropTypes.shape({
  name: PropTypes.string,
  type: PropTypes.string,
  uid: PropTypes.string,
  gatewayHid: PropTypes.string,
  softwareName: PropTypes.string,
  softwareVersion: PropTypes.string,
  hid: PropTypes.string,
  discoveredAt: PropTypes.number,
  lastPingedAt: PropTypes.number,
  timezone: PropTypes.string,
  frontButton: PropTypes.bool,
});

export const feederTelemetryShape = PropTypes.shape({
  timestamp: PropTypes.number,
  voltage: PropTypes.number,
  usb_power: PropTypes.bool,
  charging: PropTypes.bool,
  ir: PropTypes.bool,
  rssi: PropTypes.number,
});

export const feedEventShape = PropTypes.shape({
  device_name: PropTypes.string,
  device_hid: PropTypes.string,
  timestamp: PropTypes.number,
  start_time: PropTypes.number,
  end_time: PropTypes.number,
  pour: PropTypes.number,
  full: PropTypes.number,
  grams_expected: PropTypes.number,
  grams_actual: PropTypes.number,
  hopper_start: PropTypes.number,
  hopper_end: PropTypes.number,
  source: PropTypes.number,
  fail: PropTypes.bool,
  trip: PropTypes.bool,
  lrg: PropTypes.bool,
  vol: PropTypes.bool,
  bowl: PropTypes.bool,
  recipe_id: PropTypes.string,
  error: PropTypes.string,
});

export const feedHistoryShape = PropTypes.shape({
  size: PropTypes.number,
  page: PropTypes.number,
  totalSize: PropTypes.number,
  totalPages: PropTypes.number,
  data: PropTypes.arrayOf(feedEventShape),
});
